File: app/src/main/java/com/anadolstudio/adelaide/domain/utils/BitmapUtil.java
package com.anadolstudio.adelaide.domain.utils;

import android.content.ContentResolver;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.drawable.BitmapDrawable;
import android.net.Uri;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.View;
import android.webkit.MimeTypeMap;
import android.widget.ImageView;

import com.anadolstudio.core.util.BitmapDecoder;

import java.io.IOException;
import java.io.InputStream;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import io.reactivex.Observable;

public class BitmapUtil {
    public static final String TAG = BitmapUtil.class.getName();

    public static final float TURN_LEFT_90 = -90f;
    public static final float TURN_RIGHT_90 = 90f;

    public static final int COLOR_MASK = Color.parseColor("#EFEFEF");// panelBackground, c ALFA_8 в любом случае будет черным.

    public static final Integer[] ARRAY_OF_COLORS_MASK = new Integer[]{Color.RED, Color.GREEN, Color.BLUE};
    public static final String CONTENT = "content:";

    public static Observable<Bitmap> loadBitmapFromAssets(Context context, final String path) {
        return Observable.create(emitter -> {
            try {
                InputStream inputStream = context.getAssets().open(path);
                emitter.onNext(BitmapFactory.decodeStream(inputStream));
                inputStream.close();
                emitter.onComplete();
            } catch (IOException ex) {
                emitter.onError(ex);
            }
        });
    }

    public static Bitmap centerCrop(Bitmap source) {
        if (source == null) {
            return null; // TODO ?
        }

        int h = source.getHeight();
        int w = source.getWidth();
        int minSide = Math.min(w, h);
        int maxSide = Math.max(w, h);
        int x = (w >= h) ? (maxSide / 2 - minSide / 2) : 0;
        int y = (w >= h) ? 0 : (maxSide / 2 - minSide / 2);

        return Bitmap.createBitmap(source, x, y, minSide, minSide);
    }

    public static int getXSpace(Bitmap bitmap, Bitmap bitmap1) {
        return Math.abs((bitmap1.getWidth() - bitmap.getWidth()) / 2);
    }

    public static int getYSpace(Bitmap bitmap, Bitmap bitmap1) {
        return Math.abs((bitmap1.getHeight() - bitmap.getHeight()) / 2);
    }

    public static Bitmap trim(Bitmap source) {
        int firstX = 0, firstY = 0;
        int lastX = source.getWidth();
        int lastY = source.getHeight();
        int[] pixels = new int[source.getWidth() * source.getHeight()];
        source.getPixels(pixels, 0, source.getWidth(), 0, 0, source.getWidth(), source.getHeight());

        loop:
        for (int x = 0; x < source.getWidth(); x++) {
            for (int y = 0; y < source.getHeight(); y++) {
                if (pixels[x + (y * source.getWidth())] != Color.TRANSPARENT) {
                    firstX = x;
                    break loop;
                }
            }
        }
        loop:
        for (int y = 0; y < source.getHeight(); y++) {
            for (int x = firstX; x < source.getWidth(); x++) {
                if (pixels[x + (y * source.getWidth())] != Color.TRANSPARENT) {
                    firstY = y;
                    break loop;
                }
            }
        }
        loop:
        for (int x = source.getWidth() - 1; x >= firstX; x--) {
            for (int y = source.getHeight() - 1; y >= firstY; y--) {
                if (pixels[x + (y * source.getWidth())] != Color.TRANSPARENT) {
                    lastX = x;
                    break loop;
                }
            }
        }
        loop:
        for (int y = source.getHeight() - 1; y >= firstY; y--) {
            for (int x = source.getWidth() - 1; x >= firstX; x--) {
                if (pixels[x + (y * source.getWidth())] != Color.TRANSPARENT) {
                    lastY = y;
                    break loop;
                }
            }
        }
        return Bitmap.createBitmap(source, firstX, firstY, lastX - firstX, lastY - firstY);
    }

    public static int[] getBounds(Bitmap source) {
        int[] pixels = new int[source.getWidth() * source.getHeight()];
        source.getPixels(pixels, 0, source.getWidth(), 0, 0, source.getWidth(), source.getHeight());

        return getBounds(pixels, source.getWidth(), source.getHeight());
    }

    public static int[] getBounds(int[] pixels, int width, int height) {
        int firstX = 0, firstY = 0;
        int lastX = width;
        int lastY = height;

        loop:
        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                if (pixels[x + (y * width)] != Color.TRANSPARENT) {
                    firstX = x;
                    break loop;
                }
            }
        }
        loop:
        for (int y = 0; y < height; y++) {
            for (int x = firstX; x < width; x++) {
                if (pixels[x + (y * width)] != Color.TRANSPARENT) {
                    firstY = y;
                    break loop;
                }
            }
        }
        loop:
        for (int x = width - 1; x >= firstX; x--) {
            for (int y = height - 1; y >= firstY; y--) {
                if (pixels[x + (y * width)] != Color.TRANSPARENT) {
                    lastX = x;
                    break loop;
                }
            }
        }
        loop:
        for (int y = height - 1; y >= firstY; y--) {
            for (int x = width - 1; x >= firstX; x--) {
                if (pixels[x + (y * width)] != Color.TRANSPARENT) {
                    lastY = y;
                    break loop;
                }
            }
        }
        return new int[]{firstX, firstY, lastX, lastY};
    }

    public static Bitmap cropFromSource(
            int height, int weight,
            int x, int y,
            Bitmap source) {

        return Bitmap.createBitmap(
                source,
                x, y,
                weight, height);
    }

    public static Bitmap scaleBitmap(Bitmap main, Bitmap support) {
        return scaleBitmap(main.getWidth(), main.getHeight(), support, true);
    }

    public static Bitmap scaleBitmap(Bitmap main, Bitmap support, boolean isHard) {
        return scaleBitmap(main.getWidth(), main.getHeight(), support, isHard);
    }

    public static Bitmap scaleBitmap(float mainW, float mainH, Bitmap support) {
        return scaleBitmap(mainW, mainH, support, true);
    }

    public static Bitmap scaleBitmap(float mainW, float mainH, Bitmap support, boolean isHard) {
        float supportW = support.getWidth();
        float supportH = support.getHeight();
        float scaleRatio = getScaleRatio(mainW, mainH, supportW, supportH);

        int scaleW = (int) (isHard ? mainW : supportW * scaleRatio);
        int scaleH = (int) (isHard ? mainH : supportH * scaleRatio);

        return Bitmap.createScaledBitmap(support, scaleW, scaleH, true);
    }

    public static float getScaleRatio(float mainW, float mainH, float supportW, float supportH) {
        return (supportW > mainW && supportH > mainH) ?
                Math.min(mainW / supportW, mainH / supportH) :
                Math.max(mainW / supportW, mainH / supportH);
    }

    public static float getScaleRatio(float current, float defaultInt) {
        return defaultInt / current;
    }

    public static void fitImageToEdge(int widthContainer, int heightContainer, Bitmap source, ImageView scaledImageView) {
        if (source == null) {
            return;
        }

        float ratio = Math.min(
                getScaleRatio(source.getHeight(), heightContainer),
                getScaleRatio(source.getWidth(), widthContainer));

        float realWidthBitmap = source.getWidth() * (ratio >= 1 ? 1 / ratio : ratio);
        float realHeightBitmap = source.getHeight() * (ratio >= 1 ? 1 / ratio : ratio);

        float scale = Math.max(
                widthContainer / realWidthBitmap,
                heightContainer / realHeightBitmap);

        scaledImageView.setScaleX(scale);
        scaledImageView.setScaleY(scale);
    }

    public static Bitmap captureView(View view) {
        return captureView(view.getWidth(), view.getHeight(), view);
    }

    public static Bitmap captureView(int width, int height, View... views) {
        Bitmap bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
        if (views == null) {
            return bitmap;
        }

        Canvas canvas = new Canvas(bitmap);
        for (View v : views) {
            v.draw(canvas);
        }

        return bitmap;
    }

    public static Bitmap captureViewMask(View view) {
        int width = view.getWidth();
        int height = view.getHeight();

        Bitmap bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);

        Canvas canvas = new Canvas(bitmap);
        canvas.drawColor(COLOR_MASK);
        view.draw(canvas);

        return bitmap;
    }


    private static int getCorrectColor(int color) {
        int r = getColorIfThisMax(Color.red(color));
        int g = getColorIfThisMax(Color.green(color));
        int b = getColorIfThisMax(Color.blue(color));
        return Color.argb(255, r, g, b);
    }

    private static int getColorIfThisMax(int color) {
        return color == 255 ? color : 0;
    }

    @NonNull
    public static DisplayMetrics getDefaultSize(AppCompatActivity activity) {
        DisplayMetrics displayMetrics = new DisplayMetrics();
        activity.getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
        return displayMetrics;
    }

    @NonNull

    public static DisplayMetrics getRealSize(AppCompatActivity activity) {
        DisplayMetrics displayMetrics = new DisplayMetrics();
        activity.getWindowManager().getDefaultDisplay().getRealMetrics(displayMetrics);
        return displayMetrics;
    }

    public static Bitmap getBitmapFromImageView(ImageView source) {
        BitmapDrawable bitmap = ((BitmapDrawable) source.getDrawable());
        return bitmap == null ? null : bitmap.getBitmap();
    }

    public static int dpToPx(Context context, int dp) {
        DisplayMetrics displayMetrics = context.getResources().getDisplayMetrics();
        return Math.round(dp * (displayMetrics.xdpi / DisplayMetrics.DENSITY_DEFAULT));
    }

    public static int pxToDp(Context context, int px) {
        DisplayMetrics displayMetrics = context.getResources().getDisplayMetrics();
        return Math.round(px / (displayMetrics.xdpi / DisplayMetrics.DENSITY_DEFAULT));
    }

    public static void getInfoOfBitmap(Bitmap bitmap) {
        Log.d(TAG, String.format("bitmap size = %sx%s, byteCount = %s, total = %s",
                bitmap.getWidth(), bitmap.getHeight(),
                (bitmap.getByteCount()), bitmap.getHeight() * bitmap.getWidth()));
        logMemory();
    }

    public static String getMimeType(AppCompatActivity activity, Uri uri) {
        String mimeType;

        if (uri.getScheme().equals(ContentResolver.SCHEME_CONTENT)) {
            ContentResolver cr = activity.getContentResolver();
            mimeType = cr.getType(uri);
        } else {
            String fileExtension = MimeTypeMap.getFileExtensionFromUrl(uri.toString());
            mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtension.toLowerCase());
        }
        return mimeType;
    }

    public static Bitmap decodeBitmapFromPath(Context context, String path,
                                              int reqWidth, int reqHeight) {
        BitmapDecoder decoder;

        if (path.contains(CONTENT)) {
            decoder = new BitmapDecoder.FromContentPath(context);
        } else {//RealPath
            decoder = new BitmapDecoder.FromRealPath();
        }

        return decoder.decode(path, reqWidth, reqHeight);
    }

    private static void logMemory() {
        Log.i(TAG, String.format("Used memory = %s",
                (int) ((Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()) / 1024)));
        Log.i(TAG, String.format("Total memory = %s",
                (int) ((Runtime.getRuntime().totalMemory()) / 1024)));
    }
}

File: photoeditorprocessor/src/main/java/com/anadolstudio/photoeditorprocessor/util/BitmapUtil.kt
package com.anadolstudio.photoeditorprocessor.util

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.net.Uri
import androidx.exifinterface.media.ExifInterface
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

object BitmapUtil {
    const val MIME_TYPE = "image/*"
    const val MAX_SIDE = 2560
    const val MAX_SIDE_COPY = MAX_SIDE / 2

    private fun getDegree(orientation: Int) = when (orientation) {
        ExifInterface.ORIENTATION_ROTATE_90 -> 90
        ExifInterface.ORIENTATION_ROTATE_180 -> 180
        ExifInterface.ORIENTATION_ROTATE_270 -> 270
        else -> 0
    }

    fun validateUri(context: Context, path: String): Boolean {
        return context.contentResolver.openFileDescriptor(Uri.parse(path), "r").use {
            val options = BitmapFactory.Options()
            options.inJustDecodeBounds = false
            BitmapFactory.decodeFileDescriptor(it?.fileDescriptor, null, options) != null
        }
    }

    fun decodeBitmapFromContentResolverPath(
        context: Context,
        path: String,
        minSide: Int = MAX_SIDE
    ): Bitmap {

        // Читаем с inJustDecodeBounds = true для определения размеров
        val options = BitmapFactory.Options()
        var orientation = 1

        options.inJustDecodeBounds = true

        context.contentResolver.openFileDescriptor(Uri.parse(path), "r").use { pfd ->
            pfd?.let {
                BitmapFactory.decodeFileDescriptor(pfd.fileDescriptor, null, options)
                orientation = ExifInterface(pfd.fileDescriptor).getAttributeInt(
                    ExifInterface.TAG_ORIENTATION,
                    1
                )
            }
        }

        val degree = getDegree(orientation)

        options.inSampleSize = calculateInSampleSize(options, minSide)
        options.inJustDecodeBounds = false
        var bitmap: Bitmap? = null

        context.contentResolver.openFileDescriptor(Uri.parse(path), "r").use { pfd ->
            bitmap = pfd?.let {
                BitmapFactory.decodeFileDescriptor(pfd.fileDescriptor, null, options)
            } ?: throw IllegalArgumentException("PFD from path $path is null ")
        }

        if (bitmap == null) throw IllegalArgumentException("Bitmap is null")

        bitmap = scaleBitmap(bitmap!!)
        bitmap = rotate(bitmap!!, degree.toFloat())


        return bitmap!!
    }

    fun scaleBitmap(src: Bitmap, ratio: Float = MAX_SIDE.toFloat()): Bitmap {
        val srcW = src.width.toFloat()
        val srcH = src.height.toFloat()
        val scaleRatio = getScaleRatioMin(ratio, srcW, srcH)
        val scaleW = (srcW * scaleRatio).toInt()
        val scaleH = (srcH * scaleRatio).toInt()

        return Bitmap.createScaledBitmap(src, scaleW, scaleH, true)
    }

    fun getScaleRatioMin(maxSide: Float, supportW: Float, supportH: Float): Float =
        getScaleRatioMin(maxSide, maxSide, supportW, supportH)

    fun getScaleRatioMin(mainW: Float, mainH: Float, supportW: Float, supportH: Float): Float =
        if (supportW > mainW || supportH > mainH)
            min(mainW / supportW, mainH / supportH)
        else 1F

    fun getScaleRatio(current: Float, defaultInt: Float): Float = defaultInt / current

    fun getScaleRatioMax(mainW: Float, mainH: Float, supportW: Float, supportH: Float) =
        max(mainW / supportW, mainH / supportH)

    fun getScaleRatioMax(mainW: Int, mainH: Int, supportW: Int, supportH: Int) =
        getScaleRatioMax(mainW.toFloat(), mainH.toFloat(), supportW.toFloat(), supportH.toFloat())

    fun getScaleRatio(mainW: Float, mainH: Float, supportW: Float, supportH: Float): Float =
        if (supportW > mainW && supportH > mainH)
            min(mainW / supportW, mainH / supportH)
        else
            max(mainW / supportW, mainH / supportH)

    fun calculateInSampleSize(
        options: BitmapFactory.Options, maxSide: Int = MAX_SIDE
    ): Int = calculateInSampleSize(options.outWidth, options.outHeight, maxSide)

    fun calculateInSampleSize(
        width: Int, height: Int, maxSide: Int = MAX_SIDE
    ): Int {
        // Реальные размеры изображения
        var inSampleSize = 1

        if (width < maxSide && height < maxSide) return inSampleSize

        if (width > maxSide || height > maxSide) {
            val halfHeight = height / 2
            val halfWidth = width / 2

            while (halfHeight / inSampleSize > maxSide || halfWidth / inSampleSize > maxSide) {
                inSampleSize *= 2
            }
        }

        return inSampleSize
    }

    fun rotate(source: Bitmap, degree: Float): Bitmap {
        if (degree == 0f) return source
        //Если хочешь кастомный градус, то используй editor().setScaleType(ImageView.ScaleType.MATRIX);
        val matrix = createRotateMatrix(degree)
        return Bitmap.createBitmap(
            source, 0, 0,
            source.width, source.height,
            matrix, true
        )
    }

    fun createRotateMatrix(degree: Float): Matrix {
        val matrix = Matrix()
        matrix.postRotate(degree)
        return matrix
    }

    fun flip(source: Bitmap, flipHorizontally: Boolean, flipVertically: Boolean): Bitmap? {
        val matrix: Matrix = createFlipMatrix(
            source.width / 2F,
            source.height / 2F,
            flipHorizontally,
            flipVertically
        )

        return Bitmap.createBitmap(
            source, 0, 0,
            source.width, source.height,
            matrix, true
        )
    }

    fun createFlipMatrix(
        centerX: Float,
        centerY: Float,
        flipHorizontally: Boolean,
        flipVertically: Boolean
    ): Matrix {
        val matrix = Matrix()
        val xFlip = if (flipHorizontally) 1 else -1
        val yFlip = if (flipVertically) 1 else -1
        matrix.postScale(xFlip.toFloat(), yFlip.toFloat(), centerX, centerY)
        return matrix
    }

    fun getXSpace(one: Bitmap, two: Bitmap) = abs((two.width - one.width) / 2)

    fun getYSpace(one: Bitmap, two: Bitmap) = abs((two.height - one.height) / 2)

    fun cropFromSource(height: Int, weight: Int, x: Int, y: Int, source: Bitmap): Bitmap =
        Bitmap.createBitmap(source, x, y, weight, height)

    fun scaleBitmap(main: Bitmap, support: Bitmap): Bitmap =
        scaleBitmap(main.width.toFloat(), main.height.toFloat(), support, true)

    fun scaleBitmap(mainW: Float, mainH: Float, support: Bitmap): Bitmap =
        scaleBitmap(mainW, mainH, support, true)

    fun scaleBitmap(main: Bitmap, support: Bitmap, isHard: Boolean): Bitmap =
        scaleBitmap(main.width.toFloat(), main.height.toFloat(), support, isHard)

    fun scaleBitmap(mainW: Float, mainH: Float, support: Bitmap, isHard: Boolean): Bitmap =
        with(support) {
            val scaleRatio = getScaleRatio(
                mainW, mainH,
                width.toFloat(), height.toFloat()
            )

            val scaleW = (if (isHard) mainW else width.toFloat() * scaleRatio).toInt()
            val scaleH = (if (isHard) mainH else height.toFloat() * scaleRatio).toInt()

            Bitmap.createScaledBitmap(support, scaleW, scaleH, true)
        }

    fun centerCrop(source: Bitmap): Bitmap = with(source) {
        val minSide = min(width, height)
        val maxSide = max(width, height)
        val x = if (width >= height) maxSide / 2 - minSide / 2 else 0
        val y = if (width >= height) 0 else maxSide / 2 - minSide / 2
        Bitmap.createBitmap(source, x, y, minSide, minSide)
    }

}